src/DTMFSender.ts

```diff
@@ -0,0 +1,113 @@
+class DTMFSender {
+  private _outputNode: MediaStreamAudioDestinationNode;
+  private _outputStream: MediaStream;
+  private _gain: GainNode;
+  private _filter: BiquadFilterNode;
+  private _sourceNode: MediaStreamAudioSourceNode;
+  private _osc1: OscillatorNode;
+  private _osc2: OscillatorNode;
+  private _toneDuration: number = 1500;
+  private _toneGap: number = 70;
+  private _tone: string = "";
+  private _playing: boolean = false;
+
+  private _dtmfFreq: Map<string, Array<number>> = new Map([
+    ["1", [1209, 697]],
+    ["2", [1336, 697]],
+    ["3", [1477, 697]],
+    ["4", [1209, 770]],
+    ["5", [1336, 770]],
+    ["6", [1477, 770]],
+    ["7", [1209, 852]],
+    ["8", [1336, 852]],
+    ["9", [1477, 852]],
+    ["*", [1209, 941]],
+    ["0", [1336, 941]],
+    ["#", [1477, 941]]
+  ]);
+
+  constructor(sender: RTCRtpSender) {
+    var audioCtx = new AudioContext();
+    this._outputNode = audioCtx.createMediaStreamDestination();
+    this._outputStream = this._outputNode.stream;
+
+    if (sender.track) {
+      var inputStream: any = new MediaStream([sender.track]);
+      var rtpSender: any = sender;
+    } else {
+      var inputStream: any = sender;
+      this._outputStream = this._outputStream;
+      var rtpSender: any = null;
+    }
```

Comment:
I'm a little confused by some of this. `audioCtx.createMediaStreamSource` below takes a `MediaStream` as a parameter, which is fine for the `if` case, but it looks like for the `else` case this is setting `inputStream` to a `RTCRtpSender` which seems iffy. Are there active cases where `sender.track` is undefined, exercising the `else` path? I guess I'm just a little surprised that `audioCtx.createMediaStreamSource` would work when passing in a `RTCRtpSender` instead of a `MediaStream`, but if it works it works.

Reply:
I couldn't remember why I had this in there. I added this in case a `MediaStream` was passed in directly and  so we could inject audio directly into it without needing to create a new `MediaStream` wrapping the track. Not required for now, so I'll take it out.

Reply:
`audioCtx.createMediaStreamSource()` would have been called with a `MediaStream` regardless as I described above so the rest of the code would have worked - just not necessary at this time.